pkg/sdn/plugin/networkpolicy.go

```diff
@@ -535,51 +453,134 @@ func namespaceIsIsolated(ns *kapi.Namespace) bool {
 	}
 }
 
-func (np *networkPolicyPlugin) watchNamespaces() {
-	RunEventQueue(np.node.kClient.Core().RESTClient(), Namespaces, func(delta cache.Delta) error {
-		ns := delta.Object.(*kapi.Namespace)
-
-		glog.V(5).Infof("Watch %s event for Namespace %q", delta.Type, ns.Name)
-		switch delta.Type {
-		case cache.Sync, cache.Added, cache.Updated:
-			// Don't grab the lock yet since this may block
-			vnid, err := np.vnids.WaitAndGetVNID(ns.Name)
+func (np *networkPolicyPlugin) watchPods() {
+	podInformer := np.kubeInformers.Core().InternalVersion().Pods()
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*kapi.Pod)
+			np.handleAddOrUpdatePod(pod, watch.Added)
+		},
+		UpdateFunc: func(_, cur interface{}) {
+			pod := cur.(*kapi.Pod)
+			np.handleAddOrUpdatePod(pod, watch.Modified)
+		},
+		DeleteFunc: func(obj interface{}) {
+			obj, err := getDeletedObjFromInformer(obj, Pods)
 			if err != nil {
-				return err
+				glog.Error(err)
+				return
 			}
+			pod := obj.(*kapi.Pod)
+			np.handleDeletePod(pod)
+		},
+	})
+}
 
-			np.lock.Lock()
-			defer np.lock.Unlock()
-			np.kNamespaces[ns.Name] = *ns
-			if npns, exists := np.namespaces[vnid]; exists {
-				npns.isolated = namespaceIsIsolated(ns)
-				np.syncNamespace(npns)
+func (np *networkPolicyPlugin) handleAddOrUpdatePod(pod *kapi.Pod, eventType watch.EventType) {
+	glog.V(5).Infof("Watch %s event for Pod '%s/%s'", eventType, pod.Namespace, pod.Name)
+
+	// We don't want to grab np.Lock for every Pod.Status change...
+	// But it's safe to look up oldPod without locking here because no other
+	// threads modify this map.
+	oldPod, podExisted := np.pods[pod.UID]
+	if (podExisted && oldPod.Status.PodIP == pod.Status.PodIP && reflect.DeepEqual(oldPod.Labels, pod.Labels)) ||
+		(pod.Status.PodIP == "") {
```

Comment:
I realized deleting pod in this case is not a good idea. Pod network plugin setup is only called for pods with HostNetwork=false but watch pods do get pods that has HostNetwork=true and podIP is not set in this case. Since HostNetwork check was missing in watch pods, we could have deleted a valid pod. I will add HostNetwork check in watch pods and will log a warning if podIP is not set.